### PR TITLE
fix adding a circular string geometry on a compound curve layer

### LINKED ISSUE #53178: Cannot add a circular string geometry on a compound curve layer
### What is the bug or the crash?

In PyQGIS, adding a feature with a circular string geometry on a compound curve layer will fail.
This should be handled in `QgsVectorDataProvider::convertToProviderType`
https://github.com/qgis/QGIS/blob/4c6018de9238d269087b1e32e3050dbfe193f0a1/src/core/vector/qgsvectordataprovider.cpp#L897

this is not limited to PyQGIS though

### Steps to reproduce the issue

...

### Versions

master

### Supported QGIS version

- [ ] I'm running a supported QGIS version according to the roadmap.

### New profile

- [ ] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by 3nids**
fixes #53178

work in progress, I'll write some tests for this and other scenarios.

### CODE REVIEW

File: src/core/vector/qgsvectordataprovider.cpp [hunk 1]
```diff
@@ -936,6 +936,18 @@ QgsGeometry QgsVectorDataProvider::convertToProviderType( const QgsGeometry &geo
     }
   }
 
+  //convert circularstring to compoundcurve
+  if ( QgsWkbTypes::flatType( providerGeomType ) == Qgis::WkbType::CompoundCurve )
+  {
+    QgsCircularString *circularString = qgsgeometry_cast<QgsCircularString *>( geometry );
+    if ( circularString )
+    {
+      std::unique_ptr<QgsCompoundCurve> g = std::make_unique<QgsCompoundCurve>();;
```
Comment (rouault) [inline src/core/vector/qgsvectordataprovider.cpp:945]:
```suggestion
      std::unique_ptr<QgsCompoundCurve> g = std::make_unique<QgsCompoundCurve>();
```

File: src/core/vector/qgsvectordataprovider.cpp [hunk 2]
```diff
@@ -936,6 +936,18 @@ QgsGeometry QgsVectorDataProvider::convertToProviderType( const QgsGeometry &geo
     }
   }
 
+  //convert circularstring to compoundcurve
+  if ( QgsWkbTypes::flatType( providerGeomType ) == Qgis::WkbType::CompoundCurve )
+  {
+    QgsCircularString *circularString = qgsgeometry_cast<QgsCircularString *>( geometry );
+    if ( circularString )
```
Comment (nyalldawson) [inline src/core/vector/qgsvectordataprovider.cpp:943]: Should we also handle linestring -> compound curve auto translation here? Or is that done elsewhere?

**Review by lbartoletti — COMMENTED** [review]
LGTM with tests :)

**Comment by 3nids** [comment]
I was looking at this.
There is a bit of redudancy with `QgsGeometry::coerceToType`, I'm looking if we can move this,

**Comment by rouault** [comment]
The issue with coerceToType() is that it is sometimes real coercitiong, like transforming a polygon to a line, or the reverse, which is probably you don't want to do behind the back of the user. Most other conversions are OK though
It would seem reasonable to me that QgsVectorLayerEditBuffer::commitChanges() does changes that are mostly technicalities, like wrapping a circularstring in a compoudcurve and don't affect the semantics of the geometry. Perhaps coerceType() should have a "bool semanticChangeOk = true" argument that commitChanges() would set to false to forbid the line<-->polygon type of coercition?

**Comment by 3nids** [comment]
I've made a new PR for this https://github.com/qgis/QGIS/pull/53484

**Pull request closed without merging**